src/error.rs

```diff
@@ -1,103 +1,257 @@
-			},
-			ErrorInner::MissingClosingBrace { position: _ } => {
-				write!(f, "Missing closing brace")
-			},
-			ErrorInner::NoSuchVariable { position: _, name } => {
-				write!(f, "No such variable: ${}", name)
-			},
+		match self {
+			Self::InvalidEscapeSequence(e) => e.fmt(f),
+			Self::MissingVariableName(e) => e.fmt(f),
+			Self::UnexpectedCharacter(e) => e.fmt(f),
+			Self::MissingClosingBrace(e) => e.fmt(f),
+			Self::NoSuchVariable(e) => e.fmt(f),
 		}
 	}
 }
 
-impl Error {
-	/// Write source highlighting for the error location.
+/// The input string contains an invalid escape sequence.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct InvalidEscapeSequence {
+	/// The byte offset within the input where the error occurs.
 	///
-	/// The highlighting ends with a newline.
-	pub fn write_source_highlighting(&self, f: &mut impl std::fmt::Write, source: &[u8]) -> std::fmt::Result {
-		let (line, start, len) = match &self.inner {
-			ErrorInner::InvalidEscapeSequence { position, character } => {
-				let line = get_line(source, *position);
-				if character.is_some() {
-					(line, *position, 2)
+	/// This points to the associated backslash character in the source text.
+	pub position: usize,
+
+	/// The byte value of the invalid escape sequence.
+	pub character: Option<u8>,
+}
+
+impl std::error::Error for InvalidEscapeSequence {}
+
+impl std::fmt::Display for InvalidEscapeSequence {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		if let Some(c) = self.character {
+			write!(f, "Invalid escape sequence: \\{}", char::from(c))
+		} else {
+			write!(f, "Invalid escape sequence: missing escape character")
+		}
+	}
+}
+
+/// The input string contains a variable placeholder without a variable name (`"${}"`).
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct MissingVariableName {
+	/// The byte offset within the input where the error occurs.
+	///
+	/// This points to the `$` sign with a missing variable name in the input text.
+	pub position: usize,
+
+	/// The length of the variable placeholder in bytes.
+	pub len: usize,
+}
+
+impl std::error::Error for MissingVariableName {}
+
+impl std::fmt::Display for MissingVariableName {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		write!(f, "Missing variable name")
+	}
+}
+
+/// The input string contains an unexpected character.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct UnexpectedCharacter {
+	/// The byte offset within the input where the error occurs.
+	///
+	/// This points to the unexpected character in the input text.
+	pub position: usize,
+
+	/// The byte value of the unexpected character.
+	///
+	/// For multi-byte UTF-8 sequences, this only gives the value of the start byte.
+	/// You can use the `position` to get the full UTF-8 sequence from the original input string.
+	pub character: u8,
```

Comment:
Would it make sense to store a `char` in case we want to add proper unicode support later?

Ditto for `InvalidEscapeSequence`.

Reply:
Yeah, good point. Definitely more future proof. Although I want to try and actually implement proper UTF-8 support too then :p

Reply:
> Although I want to try and actually implement proper UTF-8 support too then :p

Yeah, definitely would be very useful. I was just about to report a bug related to this xD

```rust
let yaml = r"emoticon: \（ ^▽^ ）/";
if let Err(e) = substitute(yaml, &map) {
	dbg!(e.source_highlighting(yaml.as_bytes()));
}
```
```
panicked at 'byte index 12 is not a char boundary; it is inside '（' (bytes 11..14) of `emoticon: \（ ^▽^ ）/`', src\error.rs:138:18
stack backtrace:
   6: subst::error::write_underline<alloc::string::String>
             at .\src\error.rs:138
   7: subst::error::Error::write_source_highlighting<alloc::string::String>
             at .\src\error.rs:100
   8: subst::error::Error::source_highlighting
             at .\src\error.rs:109
```

Reply:
Whoops, good point :)

Reply:
Hmm, I changed `source_highlighting()` now to take a `&str` but that is not backwards compatible :(

Reply:
I'll just do a version bump I guess. Taking a `&str` is much more sensible.